---
title: "Whole Genome Nucleotide Variant Analysis for Sepi Resistance"
author: "Nicholas F. Brazeau"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  rmarkdown::html_document:
    highlight: textmate
    theme: lumen
    toc: true
    code_folding: hide
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F, fig.width = 10, fig.height = 7)
```

```{r}
#............
# imports
#..........
library(vcfR)
library(GenomicRanges)
library(tidyverse)
library(RColorBrewer)
library(grid)
library(DT)
devtools::install_github("IDEELResearch/vcfRmanip"); library(vcfRmanip)
source("~/Documents/GitHub/Sepidermidis_resistance_hosp/Ranalysis/R/00-basic_vcf_manipulations.R")

# VCF IMPORT
qcvcf <- readRDS(file = "~/Documents/GitHub/Sepidermidis_resistance_hosp/Ranalysis/data/Sepi_QC_variantfile.RDS")
# mtdt
mtdt <- readr::read_tsv(file = "~/Documents/GitHub/Sepidermidis_resistance_hosp/metadata.tab.txt") %>% 
  magrittr::set_colnames(tolower(colnames(.))) %>% 
  dplyr::mutate(collectiondate = lubridate::mdy(collectiondate)) %>% 
  dplyr::arrange(collectiondate)
```

# Background
Over their course of treatment at UNC, the patient developed multi-drug resistant _S. epidermidis_. I am currently blinded to timing and resistance phenotype data but am attempting to identify putative drug resistance mutations. **We have now performed QC and filtered the VCF and are ready for the WGS analysis**. 

## Goal 
**Overall**: To detect putative drug resistance mutation(s) in clinical isolates of _S. epidermidis_ from a patient with phenotypic resistance.   
**Immediate**: Analysis of Segregating Sites and putative Drug-rx mutations.

  
### Annotations
```{r}
# ann from vcfR snpeff
ann <- vcfR2snpeff_ann(qcvcf)

## annotation function
gffdf <- vcfRmanip::GFF2VariantAnnotation_Short("~/Documents/MountPoints/mountIDEEL/resources/genomes/Sepidermidis/info/gff/GCF_000007645.1_ASM764v1_genomic.gff")
gffdf$seqname_rn <- "Chromosome" # hacky fix to make this work with fasta and SNPeff, thankfully only one chrom
# make genomic ranges object for sepi
sepi_ranges <- GRanges(seqnames = gffdf$seqname_rn, ranges = IRanges(start = gffdf$start, end = gffdf$end), strand = factor(gffdf$strand), mcols=gffdf$GeneID) 
# make genomic ranges object for seg sites
seg_gt_sepi_ranges <- GRanges(seqnames = ann$CHROM,  ranges = IRanges(start = ann$POS, end = ann$POS)) 
# find these overlaps 
ov <- GenomicRanges::findOverlaps(sepi_ranges, seg_gt_sepi_ranges) 
gffhits <- unique(queryHits(ov)) # just the unique rows of the gff
gffdf_sub <- gffdf[gffhits,] # subset gff



gffhits.expanded <- NULL
for(i in 1:nrow(gffdf_sub)){
  temp <- NULL
  temp$POS <- seq(from=gffdf_sub$start[i], to=gffdf_sub$end[i])
  temp$CHROM <- rep(gffdf_sub$seqname_rn[i], times=length(temp$POS))
  temp$info <- rep(gffdf_sub$info[i], times=length(temp$POS))
  temp$GeneID <- rep(gffdf_sub$GeneID[i], times=length(temp$POS))
  temp$Description <- rep(gffdf_sub$Description[i], times=length(temp$POS))
  
  gffhits.expanded <- rbind(as.data.frame(gffhits.expanded), as.data.frame(temp))
   
}

Expanded.ann <- left_join(x=ann, y=gffhits.expanded, by=c("CHROM", "POS"))

DT::datatable(Expanded.ann, extensions = 'Buttons', options = list(dom = 'Bfrtip',buttons = c('csv'), pageLength=5))

# go ahead and write this out 
saveRDS(Expanded.ann, file = "~/Documents/GitHub/Sepidermidis_resistance_hosp/Ranalysis/data/Sepi_QC_annotated-variants.RDS")

```

### GT Matrix 

```{r}
DT::datatable(qcvcf@gt, extensions = 'Buttons', options = list(dom = 'Bfrtip',buttons = c('csv'), pageLength=20))

```

```{r}

chrompos <- as.tibble(t( vcfR::getFIX(qcvcf, getINFO = F) ))
chrompos$POS <- as.numeric( chrompos$POS )



plotObj <- extract.gt(qcvcf, element = "GT") %>% 
  cbind.data.frame(chrompos[,2], .) %>% 
  magrittr::set_colnames(tolower(colnames(.))) %>% 
  dplyr::mutate(pos = factor(pos)) %>% 
  tidyr::gather(data =., key = "smpl", value="gt", 2:ncol(.)) %>% 
  dplyr::mutate(smpl = factor(smpl, levels = paste0("sepi", mtdt$number))) %>% 
  ggplot(data = ., aes(x=pos, y=smpl, fill = gt)) +
  geom_tile() +
  scale_fill_brewer(palette = "Dark2") +
  theme(axis.text.x = element_blank()) 

plotly::ggplotly(plotObj)

```



# Playground
Let's turn our specificity way down and look to see if there are any other "interesting" sites. 
  

## Filter SNPs based on WSAF
Next we will look at this filtering based on the WSAF calls. 

To be very liberal, I will include all sites (no matter the quality) but will make sure they are segregating from referent genome (segregating by genotype call). Also must have at least 10x coverage at a given site for a WSAF to be considered. At least 5/8 samples must also have a "information" (e.g. FORMAT level data) at a given site for it to be considered.

```{r}
rawvcf <- vcfR::read.vcfR("~/Documents/MountPoints/mountedScratchLL/Projects/Sepidermidis_resistance_hosp/variants/Sepidermidis_resistance_hosp_HaploCaller_joint.raw.vcf")
rawvcf.segsites <- vcfRmanip::vcfR2segsites_gt(rawvcf)
ad <- vcfR::extract.gt(rawvcf.segsites, element = "AD")
dp <- vcfR::extract.gt(rawvcf.segsites, element = "DP", as.numeric = T)
filter <- dp < 10
ad[filter] <- NA


refad <- masplit(ad, record=1, sort=0, decreasing = 0)
altad <- masplit(ad, record=2, sort=0, decreasing = 0)

NRAF <- altad/(altad + refad)

# new NAs in the mix because of coverage
sites <- apply(NRAF, 1, function(x) sum(!is.na(x)))
sites <- ifelse(sites >= 5, T, F)
# subset nraf 
NRAF <- NRAF[sites, ]

chrompos <- as.tibble(vcfR::getFIX(rawvcf.segsites, getINFO = F))
chrompos$POS <- as.numeric(chrompos$POS )
chrompos <- chrompos[sites, ]


```
### Final SNP Count
**We are now left with `r nrow(NRAF)` biallelic SNP sites that have at least 10x coverage across 5/8 samples moving forward.**


## Sample Level Summary Non-Reference Allele Frequencies
### All Sites Pop AF
```{r}
apply(NRAF, 2, summary)

```

### All Sites GT Mat

```{r}
DT::datatable(NRAF[, paste0("sepi", mtdt$number)], extensions = 'Buttons', options = list(dom = 'Bfrtip',buttons = c('csv'), pageLength=20))

NRAF %>% 
  cbind.data.frame(chrompos[,2], .) %>% 
  magrittr::set_colnames(tolower(colnames(.))) %>% 
  dplyr::mutate(pos = factor(pos)) %>% 
  tidyr::gather(data =., key = "smpl", value="wsaf", 2:ncol(.)) %>% 
  dplyr::mutate(smpl = factor(smpl, levels = paste0("sepi", mtdt$number))) %>% 
  ggplot(data = ., aes(x=pos, y=smpl, fill = wsaf)) +
  geom_tile() +
  viridis::scale_fill_viridis(option="inferno") +
  theme(axis.text.x = element_blank())

```

### Subset and Viz Sites with Some Noise
Subset to sites with an WSAF standard deviation greater than 0.1.
```{r}

sddist <- NRAF %>% 
  cbind.data.frame(chrompos[,2], .) %>% 
  magrittr::set_colnames(tolower(colnames(.))) %>% 
  dplyr::mutate(pos = factor(pos)) %>% 
  tidyr::gather(data =., key = "smpl", value="wsaf", 2:ncol(.)) %>% 
  dplyr::mutate(smpl = factor(smpl, levels = paste0("sepi", mtdt$number))) %>% 
  dplyr::group_by(pos) %>% 
  dplyr::summarise(n = n(), sd = sd(wsaf, na.rm=T)) %>% 
  dplyr::filter(sd > 0.1)

plotObj <- NRAF %>% 
  cbind.data.frame(chrompos[,2], .) %>% 
  magrittr::set_colnames(tolower(colnames(.))) %>% 
  dplyr::mutate(pos = factor(pos)) %>% 
  tidyr::gather(data =., key = "smpl", value="wsaf", 2:ncol(.)) %>% 
  dplyr::mutate(smpl = factor(smpl, levels = paste0("sepi", mtdt$number))) %>% 
  dplyr::inner_join(x=., y=sddist) %>% 
  ggplot(data = ., aes(x=pos, y=smpl, fill = wsaf)) +
  geom_tile() +
  viridis::scale_fill_viridis(option="viridis") +
  theme(axis.text.x = element_blank()) 

plotly::ggplotly(plotObj)

```

**Even with these very lenient criteria, the 26858 position still looks the most interesting/informative**. 

### Table of WSAF over Time (Noisy Sites)
```{r}

NRAF[, paste0("sepi", mtdt$number)] %>% 
  DT::datatable(., extensions = 'Buttons', options = list(dom = 'Bfrtip',buttons = c('csv'), pageLength=20))
```

### Conclusions
Seems reasonable that this one SNP that passes our filters is the only clinically relevant SNP. 

