---
title: "Whole Genome INDEL Analysis for Sepi Resistance"
author: "Nicholas F. Brazeau"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  rmarkdown::html_document:
    highlight: textmate
    theme: lumen
    toc: true
    code_folding: hide
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F, fig.width = 10, fig.height = 7)

#............
# imports
#..........
library(vcfR)
library(tidyverse)
library(RColorBrewer)
library(grid)
library(DT)
devtools::install_github("IDEELResearch/vcfRmanip"); library(vcfRmanip)
source("~/Documents/GitHub/Sepidermidis_resistance_hosp/Ranalysis/R/00_functions.R")

# VCF IMPORT
lumpyvcf <- vcfR::read.vcfR("~/Documents/GitHub/Sepidermidis_resistance_hosp/lumpy/bamrefs/sepi_clincases-smoove.genotyped.vcf.gz", verbose = F)
# mtdt
mtdt <- readr::read_tsv(file = "~/Documents/GitHub/Sepidermidis_resistance_hosp/metadata.tab.txt") %>% 
  magrittr::set_colnames(tolower(colnames(.))) %>% 
  dplyr::mutate(collectiondate = lubridate::mdy(collectiondate)) %>% 
  dplyr::arrange(collectiondate)

repeatregions <- readRDS(file = "~/Documents/GitHub/Sepidermidis_resistance_hosp/Ranalysis/data/repeat_regions.RDS")

```


# Background
Over their course of treatment at UNC, the patient developed multi-drug resistant _S. epidermidis_. 

## Goal 
**Overall**: To detect putative drug resistance mutation(s) in clinical isolates of _S. epidermidis_ from a patient with phenotypic resistance.   
**Immediate**: Analysis of Segregating Sites and putative Drug-rx mutations.





```{r}

# mask lumpy

bpbuffer <- 50

repeatregions <- repeatregions %>% 
  dplyr::rename(start = start_orient,
                end = end_orient) %>% 
  dplyr::mutate(
    seqname = "Chromosome",
    geneid = 1:nrow(.),
    start = start - bpbuffer,
    end = end + bpbuffer 
  ) 


lumpyvcf.masked <- vcfRmanip::vcffilter_ChromPos(vcfRobject = lumpyvcf,
                                                 chromposbed = repeatregions)
lumpyvcf.masked.seg <- vcfRmanip::vcfR2segsites_gt(lumpyvcf.masked)



```

Of note, there appears to be a number of INDELs that are conserved across the sample as there are `r nrow(lumpyvcf.masked@gt)` loci INDELs but only `r nrow(lumpyvcf.masked.seg@gt)` segregating INDEL sites. That is a bit suprising...but may also have to do with INDEL quality. 















## Analyzing All Indels
### Locations
```{r}
chrompos <- as.tibble(vcfR::getFIX(lumpyvcf, getINFO = F))
chrompos$POS <- as.numeric(chrompos$POS )

plotObj <- vcfRmanip::gtmat012(lumpyvcf) %>%  
  cbind.data.frame(chrompos[,2], .) %>% 
  magrittr::set_colnames(tolower(colnames(.))) %>% 
  dplyr::arrange(pos) %>% 
  dplyr::mutate(pos = factor(pos)) %>% 
  tidyr::gather(data =., key = "smpl", value="gt", 2:ncol(.)) %>% 
  dplyr::mutate(gt = factor(gt)) %>% 
  dplyr::mutate(smpl = factor(smpl, levels = paste0("sepi", mtdt$number))) %>% 
  ggplot(data = ., aes(x=pos, y=smpl, fill = gt, position = pos)) +
  scale_fill_brewer(palette = "Dark2") +
  geom_tile() +
  theme(axis.text.x = element_blank()) 

plot(plotObj)


```
Maybe not so weird...does look consistent except for those hotspots...

#### Location Hotspots
```{r}
chrompos <- as.tibble(vcfR::getFIX(lumpyvcf, getINFO = F))
chrompos$POS <- as.numeric(chrompos$POS )

plotObj1 <- vcfRmanip::gtmat012(lumpyvcf) %>%  
  cbind.data.frame(chrompos[,2], .) %>% 
  magrittr::set_colnames(tolower(colnames(.))) %>% 
  dplyr::arrange(pos) %>% 
  dplyr::mutate(pos = factor(pos)) %>% 
  tidyr::gather(data =., key = "smpl", value="gt", 2:ncol(.)) %>% 
  dplyr::mutate(gt = factor(gt)) %>% 
  dplyr::mutate(smpl = factor(smpl, levels = paste0("sepi", mtdt$number))) %>% 
  ggplot(data = ., aes(x=pos, y=smpl, fill = gt, position = pos)) +
  scale_fill_brewer(palette = "Dark2") +
  geom_tile() +
  theme(axis.text.x = element_blank()) 


plotObj2 <- vcfRmanip::gtmat012(lumpyvcf) %>%  
  cbind.data.frame(chrompos[,2], .) %>% 
  magrittr::set_colnames(tolower(colnames(.))) %>% 
  dplyr::arrange(pos) %>% 
  dplyr::mutate(pos = factor(pos)) %>% 
  tidyr::gather(data =., key = "smpl", value="gt", 2:ncol(.)) %>% 
  group_by(pos) %>% 
  dplyr::summarise(gt = mean(gt)) %>% 
  dplyr::mutate(popaf = "GTmean") %>% 
  ggplot(data = ., aes(x=pos, y = popaf, fill = gt, position = pos)) +
  scale_fill_distiller("PopAF", palette = "Spectral") +
  geom_tile() +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) 


gridExtra::grid.arrange(plotObj1, plotObj2, layout_matrix = rbind(c(1), c(1), c(2)))


```



## Analyzing Segregating Site Indels
Segregating sites and at least 5/8 samples with a callable loci. 
### Locations
```{r}
chrompos <- as.tibble(vcfR::getFIX(call_sing_seg_lumpyvcf, getINFO = F))
chrompos$POS <- as.numeric(chrompos$POS )

plotObj <- vcfRmanip::gtmat012(call_sing_seg_lumpyvcf) %>%  
  cbind.data.frame(chrompos[,2], .) %>% 
  magrittr::set_colnames(tolower(colnames(.))) %>% 
  dplyr::arrange(pos) %>% 
  dplyr::mutate(pos = factor(pos)) %>% 
  tidyr::gather(data =., key = "smpl", value="gt", 2:ncol(.)) %>% 
  dplyr::mutate(gt = factor(gt)) %>% 
  dplyr::mutate(smpl = factor(smpl, levels = paste0("sepi", mtdt$number))) %>% 
  ggplot(data = ., aes(x=pos, y=smpl, fill = gt, position = pos)) +
  scale_fill_brewer(palette = "Dark2") +
  geom_tile() +
  theme(axis.text.x = element_blank()) 

plotly::ggplotly(plotObj)


```
Maybe not so weird...does look consistent except for those hotspots...

#### Location Hotspots
```{r}
chrompos <- as.tibble(vcfR::getFIX(call_sing_seg_lumpyvcf, getINFO = F))
chrompos$POS <- as.numeric(chrompos$POS )

plotObj1 <- vcfRmanip::gtmat012(call_sing_seg_lumpyvcf) %>%  
  cbind.data.frame(chrompos[,2], .) %>% 
  magrittr::set_colnames(tolower(colnames(.))) %>% 
  dplyr::arrange(pos) %>% 
  dplyr::mutate(pos = factor(pos)) %>% 
  tidyr::gather(data =., key = "smpl", value="gt", 2:ncol(.)) %>% 
  dplyr::mutate(gt = factor(gt)) %>% 
  dplyr::mutate(smpl = factor(smpl, levels = paste0("sepi", mtdt$number))) %>% 
  ggplot(data = ., aes(x=pos, y=smpl, fill = gt, position = pos)) +
  scale_fill_brewer(palette = "Dark2") +
  geom_tile() +
  theme(axis.text.x = element_blank()) 


plotObj2 <- vcfRmanip::gtmat012(call_sing_seg_lumpyvcf) %>%  
  cbind.data.frame(chrompos[,2], .) %>% 
  magrittr::set_colnames(tolower(colnames(.))) %>% 
  dplyr::arrange(pos) %>% 
  dplyr::mutate(pos = factor(pos)) %>% 
  tidyr::gather(data =., key = "smpl", value="gt", 2:ncol(.)) %>% 
  group_by(pos) %>% 
  dplyr::summarise(gt = mean(gt)) %>% 
  dplyr::mutate(popaf = "GTmean") %>% 
  ggplot(data = ., aes(x=pos, y = popaf, fill = gt, position = pos)) +
  scale_fill_distiller("PopAF", palette = "Spectral") +
  geom_tile() +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) 


gridExtra::grid.arrange(plotObj1, plotObj2, layout_matrix = rbind(c(1), c(1), c(2)))


```






